### DISCUSSION TITLE
Linked List plugin

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#403] [Show and tell]
Another plugin project: fun for me, fan-loved by me: **Linked Lists!**
Currently, it only works on Windows, where heap storage is allocated. I am investigating how this works on Linux or macOS, as equivalent methods should be available.

Interestingly, you can treat stacks as a subset of linked lists. I'll take a closer look at this.

**Up to 100 lists can be used; each function must operate on a specific list number. Named lists are not directly supported; however, this functionality can be easily implemented using REXX stem variables for list management.** 

**APPENDNODE(llist-number,data-to-add)** or **APPNODE(...)**    Add a new entry to the end of the linked list.
```
llist-number    linked-list-number 0-99 are allowed  
data-to-add     data to be stored in the linked list node
returns         0 if successfully added
```
**PREPENDNODE(llist-number,data-to-add)** or **PREPNODE(...)**   Insert a new entry at the start of the linked list.
```
llist-number    linked-list-number 0-99 are allowed  
data-to-add     data to be stored in the linked list node
returns         0 if successfully added
```
**CURRENTNODE(llist-number)**    returns the data of the current node.  
```
llist-number    linked-list-number 0-99 are allowed  
returns         data stored in the node 
```
**CURRENTNODEADDR(llist-number)** returns the storage address of the current node. 
```
 llist-number    linked-list-number 0-99 are allowed  
returns          address of the node  
```
**NEXTNODE(llist-number)**       returns the data of the next node referenced by the current node.
```
llist-number    linked-list-number 0-99 are allowed  
returns         data stored in the node 
```      
     The next node is now the current node.

**PREVNODE(llist-number)**       returns the data of the previous node referenced by the current node.
```
llist-number    linked-list-number 0-99 are allowed  
returns         data stored in the node 
```      
 The previous node is now the current node.
**SETNODE(llist-number,option)** sets the current node according to option.
```
llist-number    linked-list-number 0-99 are allowed  
option          FIRST sets to the first node
                LAST  sets to the previous node
                +n     sets the current node to the node n positions ahead (where n is an integer). If n exceeds the total number of nodes, the LAST node is used.
                -n       sets the current node to the node n positions prior (where n is an integer). If n exceeds the total number of nodes, the FIRST node is used.               
returns         address of the node if the node was successfully set
                0  if the node is not set
```
**SETNODEADDR(llist-number,node-address)** sets the current node to an existing node address.
```
llist-number    linked-list-number 0-99 are allowed  
node-address    refers to the address of an existing node, which can be obtained using CURRENTNODEADDR. Invalid addresses will cause the program to abort.
returns         address of node if the node was successfully set
                0  if the node is not set
```
**INSERTNODE(llist-number,data-to-add,"BEFORE/AFTER")** inserts a new node either prior or after the current node of the Linked List. 
```
llist-number    linked-list-number 0-99 are allowed  
data-to-add     data to be stored in the linked list node
returns         0 if successfully added
```
**LISTLLIST(llist-number)**   Print the full linked list with node pointers and data.
```
llist-number    linked-list-number 0-99 are allowed  
returns         0 if the report was successfully created
```
**REMOVENODE(llist-number)**     removes the current active node and releases its allocated memory.
 The forward and backward links of the adjacent nodes are modified accordingly.
```
llist-number    linked-list-number 0-99 are allowed  
returns         0 if the node was successfully removed
```
**FREELLIST(llist-number)**     frees an existing linked list, removes all entries, and releases all allocated memory.
```
llist-number    linked-list-number 0-99 are allowed  
returns         0 if the linked list was successfully removed
```

**Comment by rvjansen** [comment]
Looks like Simula 67's SimSet. Great!
Do note that in the CMakeLists.txt you now need to include rxfnsb instead of just rxfns due to some needed subversion I did.

**Comment by Peter-Jacob** [comment]
an improved function description and new debugging features.
# Linked List Plugin Documentation

## Overview
The llist plugin provides linked list functionality for REXX programs. It supports multiple independent linked lists, memory management, and debugging capabilities.

## Functions

### Basic Operations

#### appendnode(qname, message)  or  appnode(...) 
Appends a node to the end of the list.
- `qname`: Integer (0-99) identifying the list
- `message`: String to store in the node
- Returns: 0 on success, -8 on failure

#### prependnode(qname, message) or  prepnode(...) 
Adds a node to the beginning of the list.
- `qname`: Integer (0-99) identifying the list
- `message`: String to store in the node
- Returns: 0 on success, -8 on failure

#### insertnode(qname, message, mode)
Inserts a node relative to the current position.
- `qname`: Integer (0-99) identifying the list
- `message`: String to store in the node
- `mode`: "B" for before current node, "A" for after
- Returns: 0 on success, -8 on failure

#### removenode(qname)
Removes the current node from the list.
- `qname`: Integer (0-99) identifying the list
- Returns: 0 on success, -8 on failure

### Navigation

#### currentnode(qname)
Returns the string stored in the current node.
- `qname`: Integer (0-99) identifying the list
- Returns: String value or "$EMPTY-LLIST-n$" if empty

#### nextnode(qname)
Moves to the next node and returns its value.
- `qname`: Integer (0-99) identifying the list
- Returns: String value or "$END-OF-LLIST-n$" if at end

#### prevnode(qname)
Moves to the previous node and returns its value.
- `qname`: Integer (0-99) identifying the list
- Returns: String value or "$TOP-OF-LLIST-n$" if at start

#### setnode(qname, position)
Sets the current node position.
- `qname`: Integer (0-99) identifying the list
- `position`: "FIRST", "LAST", or relative position (+/-n)
- Returns: Address of new current node or 0

### Memory Management

#### freellist(qname)
Frees all memory associated with the list.
- `qname`: Integer (0-99) identifying the list
- Returns: 0 on success

### Debugging

#### debug(qname)
Performs memory validation and prints statistics.
- `qname`: Integer (0-99) identifying the list
- Returns: 0 if valid, -1 if corruption detected

#### cleanup()
Prints memory statistics and checks for leaks.
- Returns: 0

### Display Functions

#### listnode(qname)
Displays the contents of the list.
- `qname`: Integer (0-99) identifying the list
- Returns: 0

#### listllist(qname)
Displays detailed list information including pointers.
- `qname`: Integer (0-99) identifying the list
- Returns: 0

## Memory Protection Features

The plugin includes several memory protection features:
- Buffer overflow detection
- Memory leak detection
- Double-free detection
- Memory corruption detection
- Link integrity validation

## Example Usage
n/a
## Error and Return Codes

### Error Codes
- 0: Success - Operation completed successfully
- -8: Memory/Operation failure
  - Memory allocation failed
  - Invalid operation attempted
  - Memory free operation failed
- -1: Memory corruption detected
  - Invalid memory header magic number
  - Buffer overflow detected
  - Invalid linked list pointers

### Special Return Values
For string-returning functions:
- "$EMPTY-LLIST-n$": List n is empty (currentnode)
- "$END-OF-LLIST-n$": Reached end of list n (nextnode)
- "$TOP-OF-LLIST-n$": Reached start of list n (prevnode)
where n is the list number (qname)

For address-returning functions:
- 0: Operation failed or null pointer
- non-zero: Valid memory address

## Common Error Scenarios
1. Memory Allocation (-8)
   - System out of memory
   - Invalid size requested
   - Heap corruption

2. List Operations (-8)
   - Invalid list number (qname > 98)
   - Operation on empty list
   - Invalid current node

3. Memory Corruption (-1)
   - Buffer overflow detected
   - Invalid memory header
   - Corrupted linked list pointers
   - Double-free attempted
